repos/system_upgrade/el7toel8/actors/satelliteupgrader/actor.py

```diff
@@ -0,0 +1,27 @@
+import glob
+import os
+
+from leapp.actors import Actor
+from leapp.libraries.common.rpms import has_package
+from leapp.libraries.stdlib import run
+from leapp.models import InstalledRPM
+from leapp.tags import FirstBootPhaseTag, IPUWorkflowTag
+
+
+class SatelliteUpgrader(Actor):
+    """
+    No documentation has been provided for the satellite_upgrader actor.
+    """
+
+    name = 'satellite_upgrader'
+    consumes = (InstalledRPM, )
+    produces = ()
+    tags = (IPUWorkflowTag, FirstBootPhaseTag.After)
+
+    def process(self):
+        has_foreman = has_package(InstalledRPM, 'foreman') or has_package(InstalledRPM, 'foreman-proxy')
```

Comment:
You should rather use a custom message from your facts collection instead.

Reply:
In https://github.com/oamg/leapp-repository/pull/733/files/6d36dff7bff36d223d1adc1569accbd35d3d377e#diff-f91027c92012eac0394869e41f6b9e7c7dda170dd750466bb081d8f4de87e819 @pirat89 said not to (but back then I wasn't using the same "fact" in multiple places).

I don't mind either way, just tell me what the official convention is.

Reply:
(after some IRC convo: has_package is not really performant, and if we can save calling it multiple times with the same input, we win)

Reply:
@evgeni Hmm would it help if we had has_packages? =)

Reply:
My first thought with that: does that implement AND or OR? Would that be an option?

Reply:
Well you would still use has_package, but within the facts actor, and gather all necessary information there. 
Later actors will just consume already collected information.

e.g. if there's no message at all, then the actors can assume they don't need to do anything and just return, as there's no satellite. 

Something like this:
```python
class SatellitePostgresqlFacts(Model):
     topic = SystemFactsTopic

    same_partition = fields.Boolean(...)
    """ Whether or not target and source postgresql data will stay on the same partition """
    space_required = fields.Nullable(fields.Integer(...))
    """ How many bytes are required on the target partition """
    space_available = fields.Nullable(fields.Integer(...))
    """ How many bytes are available on the target partition """

class SatelliteFacts(Model):
     topic = SystemFactsTopic
     
    has_foreman = fields.Boolean(...)
    """Whether or not foreman is installed on this system"""
    postgresql = fields.Model(SatellitePostgresqlFacts, ...)
    """ Foreman related PostgreSQL facts """
```

Then you can do this:
```Python
self.produce(SatelliteFacts(
    has_foreman=(has_package(InstalledRPM, 'foreman') or has_package(InstalledRPM, 'foreman-proxy'),
    postgresql=SatellitePostgresqlFacts(
        same_partition=on_same_partition
        space_required=None if on_same_partition else bytes_required,
        space_available=None if on_same_partition else bytes_available,
   ),
))
```

Reply:
To group postgresql related items together. It wasn't meant to be used as a message but as a way of structuring the data. Models aren't all messages. We have been talking about this within the team already, that we want eventually mark messages as such, by having them derive from a different base class. To make clear which Models are intended to be used as messages and which are supposed to be just solely meant to structure the data.

Reply:
thanks for the explanation!